python/openinference-instrumentation/src/openinference/instrumentation/__init__.py

```diff
@@ -12,11 +33,147 @@ def suppress_tracing() -> Iterator[None]:
+        if self._metadata:
+            ctx = set_value(SpanAttributes.METADATA, self._metadata, ctx)
+        if self._tags:
+            ctx = set_value(SpanAttributes.TAG_TAGS, self._tags, ctx)
+
+        self._token = attach(ctx)
+        return
+
+    def detach_context(self) -> None:
+        detach(self._token)
+        return
+
+
+class using_attributes(AbstractContextManager, UsingAttributes):
+    """
+    TBD
+    """
+
+    slots = [
+        "_token",
+    ]
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    def __enter__(self) -> "using_attributes":
+        self.attach_context()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        self.detach_context()
+        return
+
+
+class async_using_attributes(AbstractAsyncContextManager, UsingAttributes):
+    """
+    TBD
+    """
+
+    slots = [
+        "_token",
+    ]
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    async def __aenter__(self) -> "async_using_attributes":
+        self.attach_context()
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
+        self.detach_context()
         return
-    token = attach(set_value(_SUPPRESS_INSTRUMENTATION_KEY, True))
-    yield
-    detach(token)
+
+
+def get_attributes_from_context() -> Iterator[Tuple[str, AttributeValue]]:
+    for ctx_attr in CONTEXT_ATTRIBUTES:
+        yield ctx_attr, get_value(ctx_attr)
+
+
+def is_dict_of(
+    d: Dict[object, object],
+    key_allowed_types: (Type),
+    value_allowed_types: (Type) = (),
+    value_list_allowed_types: (Type) = (),
+) -> bool:
+    """
+    Method to check types are valid for dictionary.
+
+    Arguments:
+    ----------
+        d (Dict[object, object]): dictionary itself
+        key_allowed_types (T): all allowed types for keys of dictionary
+        value_allowed_types (T): all allowed types for values of dictionary
+        value_list_allowed_types (T): if value is a list, these are the allowed types for value list
+
+    Returns:
+    --------
+        True if the data types of dictionary match the types specified by the arguments, false otherwise
+    """
+    if value_list_allowed_types and not isinstance(value_list_allowed_types, tuple):
+        value_list_allowed_types = (value_list_allowed_types,)
+
+    return (
+        isinstance(d, dict)
+        and all(isinstance(k, key_allowed_types) for k in d.keys())
+        and all(
+            isinstance(v, value_allowed_types)
+            or any(is_list_of(v, t) for t in value_list_allowed_types)
+            for v in d.values()
+            if value_allowed_types or value_list_allowed_types
+        )
+    )
+
+
+def is_list_of(lst: Sequence[object], tp: Type) -> bool:
+    return isinstance(lst, list) and all(isinstance(x, tp) for x in lst)
```

Comment:
This looks like it wants to be a typeguard.

Reply:
It is a typeguard, I forgot that we require more modern python. I was stuck on 3.6 from our SDK